### PR TITLE
Add support for multiple designated types for an operator declaration.

### PR DESCRIPTION
**Pull request by rudkx**
Add parsing, type checking, serialization, and deserialization support
for specifying multiple types as "designated" for operator lookup for
a given operator declaration.

The constraint solver still considers only the first type when
deciding the order to attempt the elements of a disjunction, so this
doesn't really change behavior yet.

### CODE REVIEW

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
It looks like I missed a few printing changes so I'll finish those up and push another commit tomorrow.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please smoke test Linux platform

**Comment by rudkx** [comment]
@swift-ci Please smoke test Linux platform

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@dmbryson Is this issue that popped up in my linux build one you're already aware of?

```
10:39:49 Failing Tests (1):
10:39:49     llbuild :: BuildSystem/Build/lane-release.llbuild
```

**Comment by rudkx** [comment]
Will retry in case this is a non-deterministic failure.

**Comment by rudkx** [comment]
@swift-ci Please smoke test Linux platform

**Comment by dmbryson** [comment]
:/ Not sure what happened there.  There should have been more detailed log messages for any of the failure conditions.

**Comment by dmbryson** [comment]
Hmm, failed again. Very strange.

**Comment by dmbryson** [comment]
Looking into it: <rdar://problem/45196629> lane-release.llbuild test is failing in some Swift-CI linux runs

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@rintaro @harlanhaskins It's not clear to me if there is a better way to deal with the libSyntax update here. This captures the syntax, but nothing about the semantic meaning of the identifiers in the list after the colon.

File: utils/gyb_syntax_support/DeclNodes.py [hunk 1]
```diff
@@ -662,31 +662,27 @@
...
-             Child('InfixOperatorGroup', kind='InfixOperatorGroup',
+             Child('OperatorPrecedenceAndTypes', kind='OperatorPrecedenceAndTypes',
                    description='''
-                   Optionally specify a precedence group
+                   Optionally specify a precedence group and designated types.
                    ''',
                    is_optional=True),
          ]),
 
+    Node('IdentifierList', kind='SyntaxCollection',
```
Comment (harlanhaskins) [inline utils/gyb_syntax_support/DeclNodes.py]: This probably needs another production that includes a trailing comma after each identifier.

File: utils/gyb_syntax_support/DeclNodes.py [hunk 2]
```diff
@@ -662,31 +662,27 @@
...
+         element='IdentifierToken'),
+
     # infix-operator-group -> ':' identifier ','? identifier?
```
[the remark this answers is not in the document]
Comment (rudkx) [reply utils/gyb_syntax_support/DeclNodes.py]: Ah yes, I'll fix that up.

File: utils/gyb_syntax_support/DeclNodes.py (hunk 1 above)
Comment (rudkx) [reply utils/gyb_syntax_support/DeclNodes.py]: I patterned this on what was happening for inheritance with `InheritedTypeList`. I don't see any notion of a comma-separated list there. Am I missing something?

**Comment by rudkx** [comment]
Latest error:

```
/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift-corelibs-libdispatch/src/swift/Block.swift:43:73: error: cannot convert value of type 'UInt32' to expected argument type 'UInt'
16:15:46                 _block =  dispatch_block_create_with_qos_class(dispatch_block_flags_t(UInt32(flags.rawValue)),
```

**Comment by rudkx** [comment]
@swift-ci Please smoke test Linux platform

**Comment by rudkx** [comment]
I'll merge the other Swift syntax fixes in a separate PR.

**Pull request merged**